### PR TITLE
[python-package] make a shallow copy on dataframe rename (fixes #4596)

### LINKED ISSUE #4596: Avoid dataframe deep copy while creating dataset
## Summary

The python API does the following while creating a dataset from a dataframe with cat cols:

```python
        if len(cat_cols):  # cat_cols is list
            data = data.copy()  # not alter origin DataFrame
            data[cat_cols] = data[cat_cols].apply(lambda x: x.cat.codes).replace({-1: np.nan})
```

By default `copy` does a deep copy (copies all data and indices) but AFAICS there is no need of that here and it could trigger a memory expensive operation.

## Motivation

Avoid extra memory usage. The sequence of steps here will end up with three copies of the data:
* the original dataframe
* the duplicated dataframe (transient)
* the dataset

## Description

Use `copy(deep=False)`:

```python
df = pd.DataFrame(dict(
    x=pd.Categorical(["a", "b", "b"]),
    y=[0, 1., 0],
))
df2 = df.copy(deep=False)
df2[["x"]] = df2[["x"]].apply(lambda x: x.cat.codes).replace({-1: np.nan})
print(df)
print(df2)
print(df2.values)
```

```
   x    y
0  a  0.0
1  b  1.0
2  b  0.0

   x    y
0  0  0.0
1  1  1.0
2  1  0.0

[[0. 0.]
 [1. 1.]
 [1. 0.]]
```

### PR DESCRIPTION
**Pull request by jmoralez**
Adds on #5225. Fixes #4596.

### CODE REVIEW

**Comment by jameslamb** [comment]
@jmoralez I want to remind you that every PR needs to have a label in https://github.com/microsoft/LightGBM/blob/59f59c9fbe060eae6114b21d6aa5f7b0ad80b4e7/.github/release-drafter.yml, so that it can be correctly categorized in the release notes.

File: tests/python_package_test/test_basic.py [hunk 1]
```diff
@@ -648,9 +648,7 @@ def test_no_copy_when_single_float_dtype_dataframe(dtype):
     pd = pytest.importorskip('pandas')
     X = np.random.rand(10, 2).astype(dtype)
     df = pd.DataFrame(X)
-    # feature names are required to not make a copy (rename makes a copy)
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py:651]: This test currently tests "when you pass a `pandas` DataFrame and explicit list of feature names, a copy is not made".

This PR is proposing changing it to "when you pass a `pandas` DataFrame and `feature_names="auto"`, a copy is not made".

To test the broader behavior "when you pass a `pandas` DataFrame, a copy is not made, regardless of how you specify feature names", would you consider instead changing the tests touched in this PR such that both cases (`feature_name` and `"auto"`) are tested?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for the very thorough investigation with profiling!

I have one question...does this change mean that `lightgbm` will now *modify* the input `pandas` DataFrame in place? Like if I bring a `pandas` DataFrame with integer column names to `lightgbm`, and I use `feature_name="auto"`, will those now be changed to strings?

If no, then I totally support this change.

If yes then I still support it (I think the memory savings is worth the potential challenges caused by silent side-effects like changing the type of column names), but think we should label this PR `breaking` and maybe include a documentation change and/or warning in that situation. I think "changing your DataFrame's column names when `lightgbm` previously used to leave it unmodified" is a breaking change that could affect users' code that works on that DataFrame post-training.

**Comment by jmoralez** [comment]
> does this change mean that lightgbm will now modify the input pandas DataFrame in place?

No, the argument `copy` to `rename` is actually passed as the `deep` argument to `DataFrame.copy`, links here: https://github.com/microsoft/LightGBM/issues/4596#issuecomment-1133784728. So it will still make a copy but a shallow one, i.e. it won't copy the underlying arrays and the index, but things like renaming the columns will only modify the copy.

**Review by jameslamb — APPROVED** [review]
Changes look great, and thanks for the answer to my question about whether the rename will have side effects outside of LightGBM.

Looks good!

**Review by StrikerRUS — APPROVED** [review]
Thanks a lot for the nice improvement and test enhancements!

**Pull request merged**